Add some optional caching

apolkosnik opened the pull request:
Also, commented out dummy cache, and 
updated some comments, and added a reference article for caching. It'll make it easier to hook CRITs up to memcached.

mgoffin commented:
What sort of stuff does this cache? Curious if we somehow leak anything we shouldn't because it got cached by a different user.

apolkosnik commented:
AFAIK this does a whole site caching. You are right, it's probably better to get a per user caching.
I was looking at https://stackoverflow.com/questions/20146741/django-per-user-view-caching 
One of the options there are:
https://gist.github.com/caot/6480c39453f5d2fa86bf
and
https://djangosnippets.org/snippets/2524/

mgoffin commented:
Ahh so we would have to define decorators and apply them to every single view. That seems cumbersome, especially if we want this to be optional and something to `enable`.

apolkosnik commented:
Hopefully it's more palatable now.

apolkosnik commented:
- adjustments to the middleware ordering for caches, and conditionals for middleware loading
- updated some comments
- uncomented the local mem cache backend, as it is in master 
- added option for turning on/off django caching (whether to load or not the middleware)

apolkosnik commented:
@mgoffin Please let me know if that's better
